src/libcore/sync/atomic.rs

```diff
@@ -1268,6 +1268,70 @@ macro_rules! atomic_int {
                 }
             }
 
+            /// fetch the value, apply a function to it and return the result
+            ///
+            /// Note: This may call the function multiple times if the value has been
+            /// changed from other threads in the meantime.
+            ///
+            /// Examples:
+            ///
+            /// ```rust
+            /// use std::sync::atömic::{AtomicIsize, Ordering};
+            ///
+            /// let x = AtomicIsize::new(0);
+            /// let mut max_x = isize.min_value();
+            /// for y in &[1, 5, 2] {
+            ///     max_x = x.update_and_fetch(|x| x.max(y));
+            /// }
+            /// ```
+            #[unstable(feature = "no_more_cas",
+                       reason = "no more CAS loops in user code"
+                       issue = "48655")]
+            #[inline]
+            fn update_and_fetch<F: Fn($int_type) -> $int_type>(&self, f: F) -> $int_type {
+                let mut prev = self.get(Ordering::Relaxed);
+                loop {
+                    let next = f(prev);
+                    match compare_exchange_weak(self, prev, next, Ordering::SeqCst, Ordering::Relaxed) {
```

Comment:
Is there any chance the user wants to use ordering other than SeqCst/Relaxed? (Similar question for the `self.get(Relaxed)` above)

Reply:
Definitely yes. Current implementation seems to be always right, however it might end up doing more loops than strictly necessary if some stricter ordering was used for `get`… I think. And `SeqCst` isn’t always necessary for CEW, but once you go for a weaker ordering there, it is important to be able to specify a stronger one for `get` and failure cases.

Reply:
@nagisa Benchmarks [in other languages](http://psy-lob-saw.blogspot.de/2015/10/expanding-queue-interface-relaxed-queue.html) show that using the most relaxed ordering and allowing for more loops achieves for better throughput. However, as this is not other languages, I'll see if I can come up with some benchmarks for it.

Reply:
@llogiq it is difficult to generalise the benchmarks like this. The article you’ve linked to is benchmarking on x86, which will have different performance characteristics compared to an architecture where the memory model is weaker. This is exactly why we provide means to specify the orderings *everywhere*.